appium/options/android/common/adb/adb_options.py

```diff
@@ -0,0 +1,383 @@
+# Licensed to the Software Freedom Conservancy (SFC) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The SFC licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from datetime import timedelta
+from typing import Any, TypeVar, Generic
+
+from appium.options.common.supports_capabilities import SupportsCapabilities
+
+T = TypeVar('T')
+C = TypeVar('C', bound='SupportsCapabilities')
+
+
+class AdbOptionsDescriptor(Generic[T]):
+    def __init__(self, name: str) -> None:
+        self.name = name
+
+    def __get__(self, obj: C, cls: type[C]) -> Any:
+        if self.name == 'ADB_EXEC_TIMEOUT':
```

Comment:
this logic must not be present in the descriptor classes, but rather incapsulated in corresponding Option classes

Reply:
> st not be present in

I tried this. I cannot have the logic for descriptor lookup in specific Option class. I have to have both `__get__` and `__set__` method in  descriptor class itself.

```
class Descriptor:
    def __get__(self, obj, cls):
        print("__get__ Descriptor")

    def __set__(self, obj, value):
        print("__set__ Descriptor")


class Option1(Descriptor):
    name = Descriptor()

    def __get__(self, obj, cls):
        print("__get__ Option1")
```
`option = Option1()`
when I do an attribute look-up `option.name` it always calls parent class `__get__` I can't override it.

Reply:
yes, the solution above won't work. What I would do is to pass two transformer functions as constructor arguments - one for the setter and one for the getter. By default these could be None, which means the value is get/set as is, otherwise it must be preprocessed by the given transformer(s) first.

Reply:
```
class AdbOptionsDescriptor(Generic[T]):
    def __init__(self, name: str, fget=None, fset=None) -> None:
        self.name = name
	self.fget = fget
        self.fset = fset

    def __get__(self, obj: C, cls: type[C]) -> Any:
        if self.fget:
            return self.fget(obj)
        return obj.get_capability(self.name)

    def __set__(self, obj: C, value: Any) -> None:
        if self.fset:
            self.fset(obj, value)
        obj.set_capability(self.name, value)


class AdbExecTimeoutOption(SupportsCapabilities):
    ADB_EXEC_TIMEOUT = 'adbExecTimeout'
    
    def getter(self):
        value = self.get_capability("ADB_EXEC_TIMEOUT")
        return None if value is None else timedelta(milliseconds=value)

    def setter(self, value):
        self.set_capability(self.name, int(value.total_seconds() * 1000) if isinstance(value, timedelta) else value)    
    
    adb_exec_timeout = AdbOptionsDescriptor('ADB_EXEC_TIMEOUT', getter, setter)



class AdbPortOption(SupportsCapabilities):
    ADB_PORT = 'adbPort'
    adb_port = AdbOptionsDescriptor('ADB_PORT')
```

In the above code we have two option classes, in the first option class `AdbExecTimeoutOption` we have defined transformation functions `getter` and `setter` and passing it to descriptor class `AdbOptionsDescriptor` as constructor argument. Now the `__get__` and `__set__` methods in the descriptor calls the transformation functions defined in `AdbExecTimeoutOption`.

In the second option class, we have not defined any transformation function. So the `__get__` and `__set__` function calls `get_capability` and `set_capability` functions directly.

Is this how you want me to extend this idea to other option class?

Reply:
Close, but not exactly that:

```python

def noop(x: T) -> T:
    return x

class OptionsDescriptor(Generic[R, V]):
    def __init__(self, name: str, tget: Optional[Callable[[Any], R]] = None, tset: Optiona[Callable[[V], Any]]) -> None:
        self.name = name
	self.tget = tget
        self.tset = tset

    def __get__(self, obj: C, cls: Type[C]) -> R:
        transformer = self.tget or noop
        return transformer(obj.get_capability(self.name))

    def __set__(self, obj: C, value: V) -> None:
        transformer = self.tset or noop
        obj.set_capability(self.name, transformer(valule))


class AdbExecTimeoutOption(SupportsCapabilities):    
    def transform_get(self, value: Union[timedelta, int]) -> Optional[timedelta]:
        return None if value is None else timedelta(milliseconds=value)

    def transform_set(self, value: Union[timedelta, int]) -> int:
        return int(value.total_seconds() * 1000) if isinstance(value, timedelta) else value
    
    adb_exec_timeout = OptionsDescriptor[Optional[timedelta], Union[Timedelta, Int]](
        'adbExecTimeout', 
        self.transform_get, 
        self.transform_set,
    )


class AdbPortOption(SupportsCapabilities):
    adb_port = OptionsDescriptor('adbPort')

```

Reply:
```
class AdbOptionsDescriptor(Generic[T]):
    def __init__(self, name: str, fget=None, fset=None) -> None:
        self.name = name
        self.fget = fget
        self.fset = fset

    def __get__(self, obj: C, cls: type[C]) -> Any:
        value = obj.get_capability(self.name)
        if self.fget:
            # transform the value and then return
            return self.fget(obj, value)
        return value

    def __set__(self, obj: C, value: Any) -> None:
        if self.fset:
            # transform the value before setting
            value = self.fset(obj, value)
        obj.set_capability(self.name, value)


class AdbExecTimeoutOption(SupportsCapabilities):
    ADB_EXEC_TIMEOUT = 'adbExecTimeout'
    
    def getter(self, value):
        return None if value is None else timedelta(milliseconds=value)

    def setter(self, value):
        return int(value.total_seconds() * 1000) if isinstance(value, timedelta) else value
    
    adb_exec_timeout = AdbOptionsDescriptor('ADB_EXEC_TIMEOUT', getter, setter)

class AdbPortOption(SupportsCapabilities):
    ADB_PORT = 'adbPort'
    adb_port = AdbOptionsDescriptor('ADB_PORT')
```

Reply:
If `fget` and `fset` is passed in, then transform the `value` and call `get_capability` and `set_capability`. Otherwise directly call `get_capability` and `set_capability` directly by passing the original `value`.  Just avoided `noop` function.  If you are fine with this, I will apply this change to all options. Thank You.

Reply:
these transformers could also be repeated for some caps (e.g. there are several values that accept and return tiemdeltas ), so you may decide extracting them into a shared module to avoid unnecessary repetition

Reply:
Exactly. I have seen lot of transformers which returns `timedeltas` which has repetitive code. Yes we can move those transformers which are repetitive into separate module. But before that I am gonna make the above changes first and then work on separate module for repetitive transformers.

Reply:
I have modified the descriptors to accept transformation functions. Moved the descriptor to separate module under `options/transformers.py` where in we have descriptor class `OptionsDescriptor`, `transform_get` and `transform_set` functions which is imported to the specific option class.  If the transformation is to specific class like in `other_options.py/CommandTimeoutsOption`, the transformation function is written in specific class  (local, cannot be re-used).